python/ray/serve/controller.py

```diff
@@ -152,39 +154,66 @@ async def _start_pending_backend_replicas(
             self, config_store: ConfigurationStore) -> None:
         """Starts the pending backend replicas in self.backend_replicas_to_start.
 
-        Starts the replica, then pushes an update to the router to add it to
-        the proper backend. If the replica has already been started, only
-        updates the router.
-
-        Clears self.backend_replicas_to_start.
+        Waits for replicas to start up, then removes them from
+        self.backend_replicas_to_start.
         """
-        replica_started_futures = []
+        fut_to_replica_info = {}
         for backend_tag, replicas_to_create in self.backend_replicas_to_start.\
                 items():
             for replica_tag in replicas_to_create:
-                replica_started_futures.append(
-                    self._start_backend_replicas(config_store, backend_tag,
-                                                 replica_tag))
-
-        # Wait on all creation task futures together.
-        await asyncio.gather(*replica_started_futures)
+                replica_handle = self._start_backend_replicas(
+                    config_store, backend_tag, replica_tag))
+                fut_to_replica_info[replica_handle.ready.remote().as_future()] = (
+                    backend_tag, replica_tag, replica_handle)
+
+        prev_warning = time.time()
+        while fut_to_replica_handle:
+            if time.time() - prev_warning > REPLICA_STARTUP_TIME_WARNING_S:
+                prev_warning = time.time()
+                logger.warning("Waited {}s for replicas to start up. Make "
+                               "sure there are enough resources to create the "
+                               "replicas.")
+
+            done, pending = await asyncio.wait(list(fut_to_replica_handle.keys()), timeout=1)
+            for fut in done:
+                backend_tag, replica_tag, replica_handle = ref_to_replica_handle.pop(fut)
+                self.backend_replicas[backend_tag][replica_tag] = replica_handle
 
         self.backend_replicas_to_start.clear()
 
     async def _start_backend_replicas(self, config_store: ConfigurationStore,
```

Comment:
Can you rename this to `start_one_backend_replica` or at least `start_backend_replica`?

Reply:
oh yeah good idea